### PR TITLE
all-gather in allocate() and add() for dense tensors

### PR DESCRIPTION
**Pull request by jcorbino-lbl**
+ Incorporated Dan's suggested improvements on `allocate()`
+ Implemented `add()` for dense tensors

### CODE REVIEW

File: src/tamm/tensor_impl.hpp [hunk 1]
```diff
@@ -1040,14 +1043,29 @@ class DenseTensorImpl: public TensorImpl<T> {
...
+      *ec_->pg().team(), t.rank,
+      [](const upcxx::global_ptr<T>& dst_buf, const upcxx::view<T>& src_buf) {
+        T*     dst = dst_buf.local();
+        size_t n   = src_buf.size();
+
+        for(size_t i = 0; i < n; ++i) dst[i] += src_buf[i];
+      },
+      upcxx::reinterpret_pointer_cast<T>(gptrs_[t.rank]) + t.offset,
+      upcxx::make_view((T*) buff_span.data(), (T*) buff_span.data() + block_size(blockid)))
+      .wait();
```
Comment (bonachea) [inline src/tamm/tensor_impl.hpp:1066]: This is a fully blocking implementation of accumulate, ie this `wait()` call stalls here for the full round-trip latency of the RPC.

Based on my understanding of [GA semantics](https://hpc.pnl.gov/globalarrays/papers/GA-UserManual-Main.pdf), the [`NGA_Acc64()`](https://hpc.pnl.gov/globalarrays/api/c_op_api.html#ACC) call below in the GA version is non-blocking and remains in-flight asynchronously until a subsequent `GA_Fence()` or `GA_Sync()` call from this rank. 

I'd really like to see us move towards more asynchronous communication (something that UPC++ supports _very_ well), to improve performance via overlap. I don't know the extent to which the latency of this routine matters for end-to-end performance, but I suspect it matters.

@jcorbino Do you have a plan to go back and remove all the `future::wait()` calls from code paths that matter to performance, or should we be designing that in starting now?

**Review by bonachea — CHANGES_REQUESTED** [review]
Added minor one request for change, and a discussion thread regarding overall strategy for asynchrony.

File: src/tamm/tensor_impl.hpp (hunk 1 above)
Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1066]: @bonachea Thanks for your quick feedback. Here's what I think:

#1 `NGA_Acc64()` is a blocking call
#2 This operation (add) should not be by default nonblocking, the calling processes expect the addition to be done before proceeding
#3 The next phase (once I'm done with completeness) is to improve performance

**Of course, we should exploit asynchrony as much as possible**, but in this particular fragment, we're just adding a buffer in a blocking way.

Comment (bonachea) [reply src/tamm/tensor_impl.hpp:1066]: `NGA_Acc64` definitely locks for _local_ completion (safe to overwrite source memory), but not remote completion (result committed to remote memory).  What you've written for UPC++ blocks for _remote_ completion, which takes much longer, and I suspect is unnecessary to fulfill the semantics of this function, thus leaving performance on the table.

Sorry I was unclear about that in my initial comment.

Comment (jcorbino-lbl) [reply src/tamm/tensor_impl.hpp:1063]: Thank you @bonachea , as we discussed via Slack, I will update this (and other) routines where asynchrony can be exploited. I will make a separate PR just for that. For now, this is good enough for completeness and consistency with respect to other parts of the code.

**Review by bonachea — APPROVED** [review]
Sounds good, thanks @jcorbino

**Pull request merged**